### PR TITLE
Python: Add shutil module sinks for path injection query

### PR DESCRIPTION
**Pull request by haby0**
If the path is controlled by some function calls of the `shutil` module, it may cause malicious attacks. For example, the `rmtree` function will delete the specified directory. The `move` function will move the specified file or move the specified file to the specified location, causing the file content to leak.

### CODE REVIEW

File: python/ql/lib/semmle/python/frameworks/Stdlib.qll [hunk 1]
```diff
@@ -2942,6 +2942,97 @@ private module StdlibPrivate {
...
+  private class ShutilCopyfileobjCall extends FileSystemAccess::Range, DataFlow::CallCfgNode {
+    ShutilCopyfileobjCall() { this = shutil().getMember("copyfileobj").getACall() }
+
+    override DataFlow::Node getAPathArgument() {
+      result in [
+          this.getArg(0), this.getArgByName("fsrc"), this.getArg(1), this.getArgByName("fdst")
+        ]
+    }
+  }
+
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/frameworks/Stdlib.qll:3010]: These arguments are not paths, but file-likes (so already opened files).

I think there is good value in keeping a note around, so we can easily remember that there is potential in modeling this function :blush: 

```suggestion
  // TODO: once we have flow summaries, model `shutil.copyfileobj` which copies the content between its' file-like arguments.
  // See https://docs.python.org/3/library/shutil.html#shutil.copyfileobj 
```

**Review by RasmusWL — CHANGES_REQUESTED** [review]
Thanks :muscle: 

Can you add the tests to the bottom of https://github.com/github/codeql/blob/cbd7434a7e5d48a16f0ae5840711ef9cabae3988/python/ql/test/library-tests/frameworks/stdlib/FileSystemAccess.py instead? Would be great to see tests using both positional and keyword arguments, like the rest of the tests :blush: 

Besides that, just a minor thing about one of the functions you modeled :blush:

File: python/ql/lib/semmle/python/frameworks/Stdlib.qll (hunk 1 above)
Comment (haby0) [reply python/ql/lib/semmle/python/frameworks/Stdlib.qll:3010]: Does `ShutilCopyfileobjCall` need to be deleted?

**Comment by haby0** [comment]
> Besides that, just a minor thing about one of the functions you modeled 😊

I don't quite understand the meaning of this sentence.

**Comment by RasmusWL** [comment]
> > Besides that, just a minor thing about one of the functions you modeled blush
> 
> I don't quite understand the meaning of this sentence.

Oh, that was just referring to the inline comment I made.

File: python/ql/lib/semmle/python/frameworks/Stdlib.qll (hunk 1 above)
Comment (RasmusWL) [reply python/ql/lib/semmle/python/frameworks/Stdlib.qll:3010]: Either that (as in my suggestion, which you can commit directly by accepting the suggestion), or by making the body of `getAPathArgument()` to be `none()`

Comment (haby0) [reply python/ql/lib/semmle/python/frameworks/Stdlib.qll]: Updated.

Comment (haby0) [reply python/ql/lib/semmle/python/frameworks/Stdlib.qll]: Do you think I can apply for a bounty next?

File: python/ql/test/library-tests/frameworks/stdlib/FileSystemAccess.py [hunk 2]
```diff
@@ -238,3 +239,30 @@ def test_fspath():
 
 tempfile.TemporaryDirectory("suffix", "prefix", "dir") # $ getAPathArgument="suffix" getAPathArgument="prefix" getAPathArgument="dir"
 tempfile.TemporaryDirectory(suffix="suffix", prefix="prefix", dir="dir") # $ getAPathArgument="suffix" getAPathArgument="prefix" getAPathArgument="dir"
+
```
Comment (RasmusWL) [inline python/ql/test/library-tests/frameworks/stdlib/FileSystemAccess.py:242]:
```suggestion

# ------------------------------------------------------------------------------
# shutil
# ------------------------------------------------------------------------------

```

**Review by RasmusWL — COMMENTED** [review]
Can you remove the entire file `python/ql/test/query-tests/Security/CWE-022-PathInjection/shutil_path_injection.py`? Since we now highlight that we model these in `python/ql/test/library-tests/frameworks/stdlib/FileSystemAccess.py`, that is good enough :+1:

**Comment by haby0** [comment]
> Can you remove the entire file `python/ql/test/query-tests/Security/CWE-022-PathInjection/shutil_path_injection.py`? Since we now highlight that we model these in `python/ql/test/library-tests/frameworks/stdlib/FileSystemAccess.py`, that is good enough 👍

No problem!

File: python/ql/test/library-tests/frameworks/stdlib/FileSystemAccess.py (hunk 2 above)
Comment (haby0) [reply python/ql/test/library-tests/frameworks/stdlib/FileSystemAccess.py:242]: Updated!

**Review by RasmusWL — APPROVED** [review]
LGTM :+1:

Will need to run tests before merging though

File: python/ql/test/library-tests/frameworks/stdlib/FileSystemAccess.py (hunk 2 above)
Comment (RasmusWL) [reply python/ql/test/library-tests/frameworks/stdlib/FileSystemAccess.py:242]: Thanks :+1:

**Comment by haby0** [comment]
> LGTM 👍
> 
> Will need to run tests before merging though


Ok. I will try to apply for the bounty next. Do you think it's okay?

**Comment by RasmusWL** [comment]
> Ok. I will try to apply for the bounty next. Do you think it's okay?

I'm not officially approving/rejecting bounties. If this additional modeling finds real vulnerabilities (that you can show), I don't see why you shouldn't be able to apply for a bounty.

**Comment by haby0** [comment]
> > Ok. I will try to apply for the bounty next. Do you think it's okay?
> 
> I'm not officially approving/rejecting bounties. If this additional modeling finds real vulnerabilities (that you can show), I don't see why you shouldn't be able to apply for a bounty.

The application has been submitted.

**Comment by RasmusWL** [comment]
We had some problems running tests on this one, but they all pass, so we can merge this now :+1:

**Comment by haby0** [comment]
> We had some problems running tests on this one, but they all pass, so we can merge this now 👍

Thanks.

**Pull request merged**